lib/network/level-net/sio-stream.js

```diff
@@ -0,0 +1,372 @@
+        if (this._destroyed) {
+            return;
+        }
+        while (this._readState.pushBuffer.length > 0) {
+            const item = this._readState.pushBuffer.shift();
+            debug('pushing item', item);
+            if (!this.push(item)) {
+                this._readState.readable = false;
+                break;
+            }
+        }
+    }
+
+    _read(size) {
+        debug(`_read(${size})`);
+        this._readState.readable = true;
+        this._pushData();
+    }
+
+    _ondata(data) {
+        debug('_ondata', this.streamId, data);
+        if (this._destroyed) {
+            return;
+        }
+        this._readState.pushBuffer.push.apply(this._readState.pushBuffer,
+                                              data);
+        if (this._readState.readable) {
+            this._pushData();
+        }
+    }
+
+    _onend() {
+        debug('_onend', this.streamId);
+        this._readState.pushBuffer.push(null);
+        if (this._readState.readable) {
+            this._pushData();
+        }
+    }
+
+    _onerror(receivedErr) {
+        debug('_onerror', this.streamId, 'error', receivedErr);
+        const err = deserializeError(receivedErr);
+        err.remote = true;
+        this.emit('error', err);
+    }
+}
+
+class SIODuplexStream extends stream.Duplex {
+    constructor(socket, streamId, maxPendingAck, ackTimeoutMs) {
+        super({ objectMode: true });
+        SIOOutputStream.prototype._initOutputStream.call(
+            this, socket, streamId, maxPendingAck, ackTimeoutMs);
+    }
+}
+SIODuplexStream.prototype._read = SIOInputStream.prototype._read;
+SIODuplexStream.prototype._write = SIOOutputStream.prototype._write;
+SIODuplexStream.prototype._writev = SIOOutputStream.prototype._writev;
+
+
+class SIOStreamSocket {
+    constructor(socket, maxPendingAck, ackTimeoutMs) {
+        assert(socket);
+        this.socket = socket;
+        this.maxPendingAck = maxPendingAck;
+        this.ackTimeoutMs = ackTimeoutMs;
+        this.remoteStreams = {};
+        this.localStreams = {};
+
+        this.socket.on('stream-data', (payload, cb) => {
+            const { streamId, data } = payload;
+            debug('received stream-data event streamId', streamId,
+                  'data', data);
+            const stream = this.remoteStreams[streamId];
+            if (!stream) {
+                debug('no remote stream registered with id', streamId);
+                return;
+            }
+            stream._ondata(data);
+            cb(null);
+        });
+
+        this.socket.on('stream-end', (payload, cb) => {
+            const { streamId } = payload;
+            debug('received stream-end event streamId', streamId);
+            const stream = this.remoteStreams[streamId];
+            if (!stream) {
+                debug('no remote stream registered with id', streamId);
+                return;
+            }
+            stream._onend();
```

Comment:
I reworked the stream event handling to make it simpler and more predictable (and correct, hopefully!)